### PR TITLE
Adapt font color to background

### PR DESCRIPTION
**Pull request by juliaroldi**
New api to adapt de font colors to darker or brighter  backgrounds.

In light mode: 

- Create a table
- Set black shade to a cell
- Type 
- The text is visible in font color white

POCAdaptiveColorsLight

In dark mode: 

- Create a table
- Set white shade to a cell
- Type 
- The text is visible in font color black

POCAdaptiveColorsDark
k

### CODE REVIEW

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Change the font color to white or some other color, so the text can be visible with a darker background
+ * @param element The element that contains text.
+ * @param newColor The font color to be applied. If not defined, the font color will be white.
+ */
+export default function adaptFontColorToBackgroundColor(
+    element: HTMLElement,
+    newFontColorInDark?: string,
+    newFontColorInBright?: string
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:14]: I didn't see when these parameters are used

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+) {
+    if (element.hasChildNodes() && element.firstElementChild?.hasAttribute('style')) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:16]: you already have null check for element.firstElementChild, so no need to check element.hasChildNodes()

Comment (JiuqingSong) [reply packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:16]: and why return if there is style?

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+const applyFontColor: Record<
+    string,
+    (element: HTMLElement, newFontColorInDark?: string, newFontColorInBright?: string) => void
+> = {
+    DARK: (element, newFontColorInDark) => (element.style.color = newFontColorInDark || '#ffffff'),
+    BRIGHT: (element, newFontColorInBright) =>
+        (element.style.color = newFontColorInBright || '#000000'),
+    DEFAULT: element => (element.style.color = ''),
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:34]: here you created 3 functions but only 1 of them will be used, the other 2 are wasted. I prefer just a regular if...else or switch...case check, no need to create those functions.

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
+const TRANSPARENT = 'transparent';
+const DARK = 'DARK';
+const BRIGHT = 'BRIGHT';
+const DEFAULT = 'DEFAULT';
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:4]: use const enum instead

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    string,
+    (element: HTMLElement, newFontColorInDark?: string, newFontColorInBright?: string) => void
+> = {
+    DARK: (element, newFontColorInDark) => (element.style.color = newFontColorInDark || '#ffffff'),
+    BRIGHT: (element, newFontColorInBright) =>
+        (element.style.color = newFontColorInBright || '#000000'),
+    DEFAULT: element => (element.style.color = ''),
+};
+
+function isBrightOrDark(color: string) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:37]: the function name makes me feel it should only return 2 possible values, but it actually can return 3

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    const lightness = calculateLightness(color);
+    if (lightness < 19) {
+        return DARK;
+    } else if (lightness > 79) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:41]: make  19, 79 constant outside, and add comment to explain why use these numbers

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+function calculateLightness(color: string) {
+    if (color === TRANSPARENT) {
+        return DEFAULT;
+    }
+    const isRGB = color.includes('rgb(');
+    const isRGBA = color.includes('rgba(');
+    let r;
+    let g;
+    let b;
+    if (isRGB || isRGBA) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:57]: better use a regular expression to parse the values.

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        b = parseInt(colors.split(',')[2]);
+    } else {
+        let colors = color.replace('#', '');
+        if (colors.length === 3) {
+            colors = colors.replace(/(.)/g, '$1$1');
+        }
+        r = parseInt(colors.substr(0, 2), 16);
+        g = parseInt(colors.substr(2, 2), 16);
+        b = parseInt(colors.substr(4, 2), 16);
+    }
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:71]: what if color is string name, such as "white", or variable, such as "--ms-color-white"?

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    r = r / 255;
+    g = g / 255;
+    b = b / 255;
+    return (Math.max(r, g, b) + Math.min(r, g, b)) * (1 / 2) * 100;
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:75]: add comment about this equation.

File: packages/roosterjs-editor-dom/lib/table/applyTableFormat.ts [hunk 10]
```diff
@@ -62,6 +63,7 @@ function setCellColor(cells: VCell[][], format: TableFormat) {
                 } else {
                     setColor(cell.td, TRANSPARENT, true /** isBackgroundColor*/);
                 }
+                adaptFontColorToBackgroundColor(cell.td);
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/table/applyTableFormat.ts:66]: I saw this function is always used after setColor is called, prefer integrate it into setColor() and add an optional parameter to specify if we need this operation, then no need to export this function

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts (hunk 1 above)
Comment (juliaroldi) [reply packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:14]: I put this parameters in case the user whats to change the font color to other color rather than black/white

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+//If the value of the lightness is more than 80, the color is bright
+const BRIGHT_COLORS_LIGHTNESS = 80;
+
+/**
+ * @internal
+ * Change the font color to white or some other color, so the text can be visible with a darker background
+ * @param element The element that contains text.
+ * @param backgroundColor The backgroundColor of a element
+ */
+export default function adaptFontColorToBackgroundColor(
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:21]: Since this function is only used by setColor(), you can move it into setColor.ts

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
+const enum ColorTones {
+    BRIGHT,
+    DARK,
+}
+const LiteralDarkColor = ['black', '--ms-color-black'];
+const LiteralBrightColor = ['white', '--ms-color-white'];
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:6]: RoosterJs should not know this kind of color variable. They are not standard.

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
+const enum ColorTones {
+    BRIGHT,
+    DARK,
+}
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:4]: If there are only two possible values, why not just use boolean?
We should only use enum when there are at least 3 values.

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+ * @param backgroundColor The backgroundColor of a element
+ */
+export default function adaptFontColorToBackgroundColor(
+    element: HTMLElement,
+    backgroundColor: string
+) {
+    if (element.firstElementChild?.hasAttribute('style')) {
+        return;
+    }
+    if (isADarkOrBrightColor(backgroundColor) === ColorTones.DARK) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:28]: No need to calculate twice.

const brightness = isADarkOrBrightColor(backgroundColor);
if (...)

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+function isADarkOrBrightColor(color: string) {
+    let lightness = 50; // set 50, because is not bright or dark
+    if (isRGB(color) || isRGBA(color) || isHEX(color)) {
+        lightness = calculateLightness(color);
+    }
+    if (lightness < DARK_COLORS_LIGHTNESS || LiteralDarkColor.indexOf(color) > -1) {
+        return ColorTones.DARK;
+    } else if (lightness > BRIGHT_COLORS_LIGHTNESS || LiteralBrightColor.indexOf(color) > -1) {
+        return ColorTones.BRIGHT;
+    }
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:46]: I think a better way is to use getComputedStyle since that will always give you rgb/rgba color (at least from what I see)

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+ * @returns
+ */
+function calculateLightness(color: string) {
+    let r; // red
+    let g; // green
+    let b; // blue
+
+    if (isRGB(color) || isRGBA(color)) {
+        //if the color representation is RGB, extract the values of red, green and blue
+        const colors = color.match(/[\d\.]+/g) as RegExpMatchArray;
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:65]: You already do isRGB or isRGBA test in line 39, no need to check again. 

Actually you can just use regex do check and parse together, no need to do isRGB/isRGBA check since regex can do that for you.

File: packages/roosterjs-editor-dom/lib/utils/setColor.ts [hunk 17]
```diff
@@ -11,19 +12,23 @@ export default function setColor(
     element: HTMLElement,
     color: string | ModeIndependentColor,
     isBackgroundColor: boolean,
-    isDarkMode?: boolean
+    isDarkMode?: boolean,
+    shouldAdaptTheFontColor?: boolean
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/setColor.ts:16]: Add parameter for shouldAdaptTheFontColor

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts (hunk 13 above)
Comment (juliaroldi) [reply packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:4]: The color can be bright or dark or neither of them, so we have three options

Comment (JiuqingSong) [reply packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:4]: But the enum ColorTones only has two values: BRIGHT and DARK.

File: packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts (hunk 15 above)
Comment (juliaroldi) [reply packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:46]: getComputedStyle will return the light gray color for when the element is selected

Comment (juliaroldi) [reply packages/roosterjs-editor-dom/lib/utils/adaptFontColorToBackgroundColor.ts:46]: Using getPropertyValue will always return RGB or RGBA

File: packages/roosterjs-editor-dom/lib/utils/setColor.ts [hunk 18]
```diff
@@ -36,3 +45,49 @@ export default function setColor(
...
+}
+
+function isADarkOrBrightColor(color: string): boolean | null {
+    let lightness = calculateLightness(color);
+    if (lightness < DARK_COLORS_LIGHTNESS) {
+        return true;
+    } else if (lightness > BRIGHT_COLORS_LIGHTNESS) {
+        return false;
+    }
+    return null;
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/setColor.ts:76]: I like the enum solution. You just need to define a 3 value enum:

const enum Darkness = {
 None,
 Light,
 Dark,
}

File: packages/roosterjs-editor-dom/lib/utils/setColor.ts [hunk 19]
```diff
@@ -36,3 +45,49 @@ export default function setColor(
...
+}
+
+/**
+ * Calculate the lightness of HSL (hue, saturation and lightness) representation
+ * @param color a RBG COLOR
+ * @returns
+ */
+function calculateLightness(color: string) {
+    let [r, g, b] = color.match(/[\d\.]+/g) as RegExpMatchArray;
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/setColor.ts:85]: the Regex should also include "rgb" and "rgba", for example:

/^rgba?\(([\d\.]+)\s*,\s*([\d\.]+)\s*,\s*([\d\.]+)(\s*,\s*[\d\.]+)?\)$/

File: packages/roosterjs-editor-dom/lib/utils/setColor.ts [hunk 20]
```diff
@@ -36,3 +45,49 @@ export default function setColor(
...
+    // Use the values of r,g,b to calculate the lightness in the HSl representation
+    //First calculate the fraction of the light in each color, since in css the value of r,g,b is in the interval of [0,255], we have
+    const red = parseInt(r) / 255;
+    const green = parseInt(g) / 255;
+    const blue = parseInt(b) / 255;
+    //Then the lightness in the HSL representation is the average between maximum fraction of r,g,b and the minimum fraction
+    return (Math.max(red, green, blue) + Math.min(red, green, blue)) * (1 / 2) * 100;
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/setColor.ts:92]: "* (1 / 2) * 100" can be " * 50", right?

File: packages-ui/roosterjs-react/lib/ribbon/component/buttons/cellShade.ts [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+    BackgroundColorDropDownItems,
+    BackgroundColors,
+    getColorPickerDropDown,
+} from './colorPicker';
+
+/**
+ * @internal
+ * "Cell Shade" button on the format ribbon
+ */
+export const cellShade: RibbonButton<CellShadeButtonStringKey> = {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/ribbon/component/buttons/cellShade.ts:14]: this button only works for table, right? So we should add isDisabled() function to check if focus is in table. You can add a "isInTable" property in ElementBasedFormatState, and set it in getElementBasedFormatState().

If that is too much change, you can do this in a separate PR, and do not add this button in getButtons() for now.

File: packages/roosterjs-editor-dom/lib/utils/setColor.ts [hunk 22]
```diff
@@ -23,7 +38,6 @@ export default function setColor(
                 ? modeIndependentColor?.darkModeColor
                 : modeIndependentColor?.lightModeColor) || colorString
         );
-
         if (element.dataset) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/setColor.ts:41]: nit, do not remove the empty line before "if", normally we should add empty line before those structural statement such as if, for, while, ...

File: packages/roosterjs-editor-dom/lib/utils/setColor.ts [hunk 23]
```diff
@@ -34,5 +48,66 @@ export default function setColor(
...
+    switch (isADarkOrBrightOrNone) {
+        case ColorTones.DARK:
+            element.style.color = WHITE;
+            break;
+        case ColorTones.BRIGHT:
+            element.dataset[DarkModeDatasetNames.OriginalStyleColor] = WHITE;
+            element.style.color = BLACK;
+            break;
+        default:
+            element.style.color = '';
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-dom/lib/utils/setColor.ts:80]: why remove color?
